File: src/python/graph/9_circuit_parser.py
# file: 9_circuit_parser.py
# Extract tagged sections and replace separators with commas

import sys # argv, exit
import re  # findall, sub, DOTALL

def main():
    if len(sys.argv) != 3: sys.exit(1)
    in_p, out_p = sys.argv[1], sys.argv[2]
    with open(in_p, 'r') as f: content = f.read()
    matches = re.findall(r'<<<(\w+)>>> START\n(.*?)<<<\1>>> END', content, re.DOTALL)
    res = []
    for tag, block in matches:
        block = re.sub(r'\s+', '', block)
        for c in '\n()-': block = block.replace(c, ',')
        res.append(f"{tag}:{re.sub(r',+', ',', block).strip(',')}")
    with open(out_p, 'w') as f: f.write('\n'.join(res))

if __name__ == "__main__":
    main()

File: src/typescript/graph/9_circuit_parser.ts
import fs from 'fs';

export function circuitParser(inputPath: string, outputPath: string): void {
  if (!fs.existsSync(inputPath)) return;
  const content = fs.readFileSync(inputPath, 'utf8');
  const matches = content.matchAll(/<<<(\w+)>>> START\n([\s\S]*?)<<<\1>>> END/g);
  const results: string[] = [];

  for (const m of matches) {
    let block = m[2].replace(/\s+/g, '');
    ['\n', '(', ')', '-'].forEach(c => block = block.split(c).join(','));
    results.push(`${m[1]}:${block.replace(/,+/g, ',').replace(/(^,|,$)/g, '')}`);
  }
  fs.writeFileSync(outputPath, results.join('\n'));
}
